opencode: 0.0.52 -> 0.1.194

delafthi opened the pull request:
### Summary

- Added myself as a maintainer for the package.
- Updated `opencode` from version `0.0.52` to `0.1.126`. This update includes several important changes:

### Details

- The project was forked due to [this issue](https://x.com/thdxr/status/1933561254481666466).
- Since most of the project was rewritten in TypeScript and built with Bun, I changed the build process to download precompiled binaries instead of building from source.
- While this is not an ideal solution, I tried to base this approach on the method used for [helix-gpt](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/he/helix-gpt/package.nix) to manage Bun packaging. However, I was unable to get the `sharp` dependency to compile, and I couldn’t find any existing references or solutions to this problem. As a result, using prebuilt binaries is the current workaround until a better build process for Bun-based packages is available.
- This PR includes breaking changes to the package. However, since the package is still in early development, I believe updating the release notes is not necessary.

### Notes

I'll also add a home-manager module -> see https://github.com/nix-community/home-manager/pull/7320

---

Please let me know if you need any additional information or testing!

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

delafthi committed:
maintainers: add delafthi

delafthi committed:
opencode: 0.0.52 -> 0.1.169

BREAKING CHANGE: This version forked from the original opencode and was
partially rebuilt. Therefore, there are various breaking changes.

repparw commented:
mind updating to v0.1.136?

GaetanLepage reviewed (commented):
Thanks for handling this non-trivial update!

GaetanLepage commented:
It's unfortunate we have to fetch the prebuilt binaries...
What  was the exact problem you encountered with `sharp`? Can you share logs?

delafthi commented:
I currently don’t have access to the logs. @thdxr, thanks for pointing that out. I'll try this and provide feedback as soon as I have an update.

delafthi commented:
> I am trying to use this derivation, but the binary is just bun. Each time I run `opencode -h` all I get is bun, `opencode --version`, 1.2.17, which is bun's version.
> 
> 
> 
> Doesn't the opencode project provide the actual project in its releases? Why would they package bun?!

I’ll try to push an update tomorrow to build opencode from source instead of just installing the binaries, so probably this issue should be resolved then.

GaetanLepage commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,60 +1,137 @@
-
+let
+  version = "0.1.162";
   src = fetchFromGitHub {
     owner = "sst";
     repo = "opencode";
-    tag = "v${finalAttrs.version}";
-    hash = "sha256-wniGu8EXOI2/sCI7gv2luQgODRdes7tt1CoJ6Gs09ig=";
+    tag = "v${version}";
+    hash = "sha256-GNF7SWMh9FZjmL3EV0pT6gsR3HeEWZ+SpNWaParRwho=";
```
Why did you move away from `finalAttrs`? It makes overriding easier.

GaetanLepage replied:
Ok my bad. Nevermind.

delafthi replied:
Both `version` and `src` are defined in a `let` expression since they are shared across all three declarations (`opencode-tui`, `opencode-node-modules`, and `opencode`). I don't know how I could solve this differently.

delafthi commented:
@GaetanLepage I also resolved your comments.

repparw commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`

---
### `x86_64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

repparw commented:
error: hash mismatch in fixed-output derivation '/nix/store/1wz3fhhx76891alvvhkjpwbpyszxq2zx-opencode-0.1.162-node-modules.drv':
         specified: sha256-WiOFWCpB9aQ4TvvOSoaVHzt6KBbEjQMjDs2LDm38/B4=
            got:    sha256-DQ259QnYp04xbnopqoXUh4b0JVXoJEUvfYIC084oH0M=
error: 1 dependencies of derivation '/nix/store/mbfjc7zp4vmw12xn6p2clc0jm4x21gps-opencode-0.1.162.drv' failed to build
error: 1 dependencies of derivation '/nix/store/1ghcywz6f6bv2zymz78dk1wjal85f92v-review-shell.drv' failed to build

I'm getting this, not sure why? Does it work on your side?

delafthi commented:
Indeed, the hash value differs across systems. I'm not sure why it worked on my aarch64-linux setup with the same hash. The latest update should resolve this issue.

delafthi commented:
The output hash of `opencode-node-modules` has just changed, causing the build to fail. I need to investigate further to determine why the build process for this package is not reproducible.

repparw commented:
don't know how this gets created in upstream, but can't we get those node modules some other way? changing 6 hashes on package upgrade at this point looks wonky haha

delafthi commented:
The output hash of `opencode-node-modules` has just changed, causing the build to fail. I need to investigate further to determine why the build process for this package is not reproducible.

repparw commented:
> I agree. Let's hope for `buildBunModule`.
seems pretty stale, sadly

repparw commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

repparw commented:
I'm getting bun

```
[nix-shell:~/.cache/nixpkgs-review/pr-419604-8]$ opencode 
Bun is a fast JavaScript runtime, package manager, bundler, and test runner. (1.2.17+282dda62c)

Usage: bun <command> [...flags] [...args]

Commands:
  run       ./my-script.ts       Execute a file with Bun
            lint                 Run a package.json script
  test                           Run unit tests with Bun
  x         eslint               Execute a package binary (CLI), installing if needed (bunx)
  repl                           Start a REPL session with Bun
  exec                           Run a shell script directly with Bun

  install                        Install dependencies for a package.json (bun i)
  add       @zarfjs/zarf         Add a dependency to package.json (bun a)
  remove    babel-core           Remove a dependency from package.json (bun rm)
  update    zod                  Update outdated dependencies
  audit                          Check installed packages for vulnerabilities
  outdated                       Display latest versions of outdated dependencies
  link      [<package>]          Register or link a local npm package
  unlink                         Unregister a local npm package
  publish                        Publish a package to the npm registry
  patch <pkg>                    Prepare a package for patching
  pm <subcommand>                Additional package management utilities
  info      tailwindcss          Display package metadata from the registry

  build     ./a.ts ./b.jsx       Bundle TypeScript & JavaScript into a single file

  init                           Start an empty Bun project from a built-in template
  create    svelte               Create a new project from a template (bun c)
  upgrade                        Upgrade to latest version of Bun.
  <command> --help               Print help text for command.

Learn more about Bun:            https://bun.sh/docs
Join our Discord community:      https://bun.sh/discord
```

delafthi commented:
> I'm getting bun
> 
> ```
> [nix-shell:~/.cache/nixpkgs-review/pr-419604-8]$ opencode 
> Bun is a fast JavaScript runtime, package manager, bundler, and test runner. (1.2.17+282dda62c)
> 
> Usage: bun <command> [...flags] [...args]
> 
> Commands:
>   run       ./my-script.ts       Execute a file with Bun
>             lint                 Run a package.json script
>   test                           Run unit tests with Bun
>   x         eslint               Execute a package binary (CLI), installing if needed (bunx)
>   repl                           Start a REPL session with Bun
>   exec                           Run a shell script directly with Bun
> 
>   install                        Install dependencies for a package.json (bun i)
>   add       @zarfjs/zarf         Add a dependency to package.json (bun a)
>   remove    babel-core           Remove a dependency from package.json (bun rm)
>   update    zod                  Update outdated dependencies
>   audit                          Check installed packages for vulnerabilities
>   outdated                       Display latest versions of outdated dependencies
>   link      [<package>]          Register or link a local npm package
>   unlink                         Unregister a local npm package
>   publish                        Publish a package to the npm registry
>   patch <pkg>                    Prepare a package for patching
>   pm <subcommand>                Additional package management utilities
>   info      tailwindcss          Display package metadata from the registry
> 
>   build     ./a.ts ./b.jsx       Bundle TypeScript & JavaScript into a single file
> 
>   init                           Start an empty Bun project from a built-in template
>   create    svelte               Create a new project from a template (bun c)
>   upgrade                        Upgrade to latest version of Bun.
>   <command> --help               Print help text for command.
> 
> Learn more about Bun:            https://bun.sh/docs
> Join our Discord community:      https://bun.sh/discord
> ```

I also ran `nixpkgs-review pr 419604` and everything works on my side. Though I'm on `aarch64-darwin`. Tomorrow I'm able to test on `x86_64-linux`.

```bash
~/.cac/nix/pr-419604-1 via  impure (review-shell)
[bsh] > opencode --help

█▀▀█ █▀▀█ █▀▀ █▀▀▄ █▀▀ █▀▀█ █▀▀▄ █▀▀
█░░█ █░░█ █▀▀ █░░█ █░░ █░░█ █░░█ █▀▀
▀▀▀▀ █▀▀▀ ▀▀▀ ▀  ▀ ▀▀▀ ▀▀▀▀ ▀▀▀  ▀▀▀

Commands:
  opencode [project]         start opencode tui                        [default]
  opencode run [message..]   run opencode with a message
  opencode auth              manage credentials
  opencode upgrade [target]  upgrade opencode to the latest or a specific
                             version
  opencode serve             starts a headless opencode server
  opencode models            list all available models

Positionals:
  project  path to start opencode in                                    [string]

Options:
      --help        show help                                          [boolean]
  -v, --version     show version number                                [boolean]
      --print-logs  print logs to stderr                               [boolean]
```

debashisbiswas commented:
I can also reproduce - I'm getting opencode when I run on `aarch64-darwin`, but bun on `x86_64-linux`.

repparw commented:
I don't see anything on the PR that could cause this, really.
Does building opencode manually work? To discard something wrong upstream

delafthi commented:
Mmh, building opencode manually works. I'm not sure what causes this issue.

debashisbiswas commented:
Before finding this PR, I created a derivation in my personal config last week that downloaded the prebuilt binary and had the same issue (the binary running bun instead of opencode). I only had this issue with the derivation, though. Downloading the release manually, unzipping it, and running it directly did not have this issue.

repparw commented:
> Before finding this PR, I created a derivation in my personal config last week that downloaded the prebuilt binary and had the same issue (the binary running bun instead of opencode). I only had this issue with the derivation, though. Downloading the release manually, unzipping it, and running it directly did not have this issue.

https://github.com/debashisbiswas/.dotfiles/commit/56a04491162cf23e59bb573f84f1387a2e186944 this the one?

delafthi commented:
For reference: https://github.com/oven-sh/bun/issues/4103 and https://discourse.nixos.org/t/add-dynamically-linked-binary-to-devshell/58655 are similar issues.
I tried using `autoPatchelfHook` without success.

repparw commented:
adapting @debashisbiswas derivation in my config I get

```
❯ opencode --help
Could not start dynamically linked executable: opencode
NixOS cannot run dynamically linked executables intended for generic
linux environments out of the box. For more information, see:
https://nix.dev/permalink/stub-ld
```

debashisbiswas commented:
> adapting @debashisbiswas derivation in my config I get
> 
> ```
> ❯ opencode --help
> Could not start dynamically linked executable: opencode
> NixOS cannot run dynamically linked executables intended for generic
> linux environments out of the box. For more information, see:
> https://nix.dev/permalink/stub-ld
> ```

Maybe relevant for this discussion: My NixOS system uses `nix-ld`. It looks like @delafthi's config does as well, but I'm not sure what system they're testing on.

delafthi commented:
> adapting @debashisbiswas derivation in my config I get
> 
> ```
> ❯ opencode --help
> Could not start dynamically linked executable: opencode
> NixOS cannot run dynamically linked executables intended for generic
> linux environments out of the box. For more information, see:
> https://nix.dev/permalink/stub-ld
> ```

Did you try using `autoPatchelfHook`?

delafthi commented:
Found a solution by using `makeBinaryWrapper`.

delafthi commented:
Update to 0.1.169

repparw commented:
```
[nix-shell:~/.cache/nixpkgs-review/pr-419604-9]$ opencode
Error: Unexpected error, check log file at /home/repparw/.local/share/opencode/log/2025-06-30T201906.log for more details
24 |     "dist/extend.*",
         ^
warn: wildcard sideEffects are not supported yet, which means this package will be deoptimized
   at /nix/store/a3vj4jd2hxrn9la7dqdfxbkc0w38qnqj-opencode-0.1.162-node-modules/node_modules/zod-openapi/package.json:24:5
```

```
INFO  2025-06-30T20:19:39 +38ms service=default version=dev args= opencode
INFO  2025-06-30T20:19:39 +0ms service=app cwd=/nix/store/j07hs2cgvq44s49lb0b37z9s28afx4ky-opencode-0.1.162 creating
INFO  2025-06-30T20:19:39 +0ms service=app git
INFO  2025-06-30T20:19:39 +1ms service=bus type=storage.write subscribing
INFO  2025-06-30T20:19:39 +0ms service=app name=bus registering service
INFO  2025-06-30T20:19:39 +0ms service=format init
INFO  2025-06-30T20:19:39 +0ms service=bus type=file.edited subscribing
INFO  2025-06-30T20:19:39 +0ms service=config.hooks init
INFO  2025-06-30T20:19:39 +0ms service=bus type=file.edited subscribing
INFO  2025-06-30T20:19:39 +0ms service=bus type=session.idle subscribing
INFO  2025-06-30T20:19:39 +0ms service=app name=provider registering service
INFO  2025-06-30T20:19:39 +1ms service=app name=config registering service
INFO  2025-06-30T20:19:39 +2ms service=config $schema=https://opencode.ai/config.json theme=system provider=[object Object] experimental=[object Object] loaded
INFO  2025-06-30T20:19:39 +1ms service=models.dev refreshing
INFO  2025-06-30T20:19:39 +0ms service=provider init
INFO  2025-06-30T20:19:39 +1ms service=provider providerID=google-cloud-code found
ERROR 2025-06-30T20:19:39 +3ms service=default name=Error message=Executable not found in $PATH: "go" fatal
```

should go be a dep?

repparw commented:
![image](https://github.com/user-attachments/assets/ab0c9c7d-9ab7-4ec8-94e8-3fd713fc3e36)
Doing nix-shell -p go inside the review shell works

delafthi commented:
@repparw could you try once more. I had `opencode-tui` commented out in commit [2d50aeb](https://github.com/NixOS/nixpkgs/pull/419604/commits/2d50aebee9d85a4d43a62259dcc42e6366ad65a3) to generate the node-module hashes faster, and forgot to uncomment it again. That should be fixed in the latest push.

delafthi commented:
> getting the same. dunno how to check what commit is actually reviewing on, but I ran it after you tagged me and it's still searching for go exec

Mmh, strange.

delafthi commented:
Also, this error message after every execution isn't too pretty

```
 > result/bin/opencode
24 |     "dist/extend.*",
         ^
warn: wildcard sideEffects are not supported yet, which means this package will be deoptimized
   at /nix/store/simnias8r4b22i81nc34ray96cyk792n-opencode-0.1.169-node-modules/node_modules/zod-openapi/package.json:24:5

```

repparw commented:
oh, my bad. didn't think to say that --help works. (pretty much all seems to work except tui, at least I tried auth and serve and they gave me something that seemed logical). I don't have a problem with that but it still needs to get merged, so, however you want

repparw commented:
> warn: wildcard sideEffects are not supported yet, which means this package will be deoptimized

https://github.com/oven-sh/bun/issues/5241

delafthi commented:
> oh, my bad. didn't think to say that --help works. (pretty much all seems to work except tui, at least I tried auth and serve and they gave me something that seemed logical). I don't have a problem with that but it still needs to get merged, so, however you want

I have pushed the solution with Go included in its path.

repparw commented:
I lost what you had done with -tui with all the rebases haha, would you mind pushing that to another branch?

I think either we should have a way of exposing both to $out, or build first the -tui and use it as an input for opencode.

Maybe as a dependency in run-time? see buildInputs vs nativeBuildInputs

delafthi commented:
> I lost what you had done with -tui with all the rebases haha, would you mind pushing that to another branch?
> 
> I think either we should have a way of exposing both to $out, or build first the -tui and use it as an input for opencode.
> 
> Maybe as a dependency in run-time? see buildInputs vs nativeBuildInputs

I agree. I could do better 😄

@alejandro-angulo thanks for linking that issue. According to https://github.com/sst/opencode/issues/462#issuecomment-3014503697 elf stripping needs to be disabled else the javascript code is removed from the bundle. So I added `dontStrip = true` to the old package derivation with `bun build`. This seems to work for me on `aarch64-linux`. Can someone confirm that?

kesor commented:
> elf stripping needs to be disabled else the javascript code is removed from the bundle and we just get bun. So I added `dontStrip = true` to the old package derivation with `bun build`. This seems to work for me on `aarch64-linux`. Can someone confirm that?

Can confirm it works on `x86_64-linux` with the stripping disabled, I get actual opencode working instead of bun.

repparw commented:
Works! 🥳

kesor commented:
Another "strangeness" with this derivation. I changed the version variable to `0.1.171` without changing anything else. It built and reports 0.1.171 with `opencode --version`, but it didn't say that any of the hashes were incorrect. idk if its just the diff of 169 vs 171 that has the same hashes, or if its the derivation that masks the version and forces the version into the binary via `-Xmain.Version` and `--define OPENCODE_VERSION`.

repparw commented:
> Another "strangeness" with this derivation. I changed the version variable to `0.1.171` without changing anything else. It built and reports 0.1.171 with `opencode --version`, but it didn't say that any of the hashes were incorrect. idk if its just the diff of 169 vs 171 that has the same hashes, or if its the derivation that masks the version and forces the version into the binary via `-Xmain.Version` and `--define OPENCODE_VERSION`.

hash is definitely different for 0.1.171, so *something's wrong*

```
❯ , nurl https://github.com/sst/opencode v0.1.171
$ nix flake prefetch --extra-experimental-features 'nix-command flakes' --json github:sst/opencode/v0.1.171
fetchFromGitHub {
  owner = "sst";
  repo = "opencode";
  rev = "v0.1.171";
  hash = "sha256-aQ9fjFko1/6DshNpNuLaJP89RGIzRgFPMRINxclDX0w=";
}
```

repparw commented:
Probably this should just be merged, after all it works, and this moved to a high-level "How to package bun programs" discussion

repparw commented:
> Don't know if yarn is easier to build w/ nixpkgs.

There's mkYarnPackage, so that would be a lot easier

We would need to get a yarn.lock committed upstream, as putting it in nixpkgs is probably not okay (@thdxr would y'all be okay with that?)

https://nixos.org/manual/nixpkgs/stable/#javascript-upstream-package-manager

delafthi commented:
> > Don't know if yarn is easier to build w/ nixpkgs.
> 
> There's mkYarnPackage, so that would be a lot easier
> 
> We would need to get a yarn.lock committed upstream, as putting it in nixpkgs is probably not okay (@thdxr would y'all be okay with that?)
> 
> [nixos.org/manual/nixpkgs/stable#javascript-upstream-package-manager](https://nixos.org/manual/nixpkgs/stable/#javascript-upstream-package-manager)

The `mkYarnPackage` function is planned to be deprecated in the near future (see https://github.com/NixOS/nixpkgs/issues/324246). Therefore, we should transition to using `yarnConfigHook` and `yarnBuildHook` instead. 

I have started updating the derivation accordingly, but a working solution is not ready yet. For now, I will generate a `yarn.lock` file and include it in `nixpkgs`.

repparw commented:
> > @GaetanLepage would you try reviewing this [PR from helix-gpt](https://github.com/NixOS/nixpkgs/pull/354767) with the sandbox? I'm checking the PR's and I don't think helix-gpt was ever tested like that, so maybe it's merged like that
> 
> Sure:
> ```
> ❯ nom build github:NixOS/nixpkgs/df67f3f7b25a86dbc6e02399a6860e5a68410d87#helix-gpt --rebuild
> helix-gpt> checking outputs of '/nix/store/z8d9k6fzax1ksv7phi23g6fg8ki3clrm-helix-gpt-0.34.drv'
> helix-gpt> Running phase: unpackPhase
> helix-gpt> unpacking source archive /nix/store/jnllq92xgrbfiyfnj9q4czlqkfi0ix68-source
> helix-gpt> source root is source
> helix-gpt> Running phase: patchPhase
> helix-gpt> Running phase: updateAutotoolsGnuConfigScriptsPhase
> helix-gpt> Running phase: installPhase
> helix-gpt> Running phase: fixupPhase
> helix-gpt> shrinking RPATHs of ELF executables and libraries in /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34
> helix-gpt> shrinking /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34/bin/helix-gpt
> helix-gpt> checking for references to /build/ in /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34...
> helix-gpt> patching script interpreter paths in /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34
> helix-gpt> stripping (with command strip and flags -S -p) in  /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34/bin
> ┏━ Dependency Graph:
> ┃ ✔ helix-gpt-0.34
> ┣━━━ Builds
> ┗━ ∑ ⏵ 0 │ ✔ 1 │ ⏸ 0 │ Finished at 10:24:26 after 0s
> ```

that does not seem to use sandbox as nixpkgs-review did, I was expecting the same as here (pass on native Darwin, fail on sandbox)

repparw commented:
> > > > @GaetanLepage would you try reviewing this [PR from helix-gpt](https://github.com/NixOS/nixpkgs/pull/354767) with the sandbox? I'm checking the PR's and I don't think helix-gpt was ever tested like that, so maybe it's merged like that
> > > 
> > > 
> > > Sure:
> > > ```
> > > ❯ nom build github:NixOS/nixpkgs/df67f3f7b25a86dbc6e02399a6860e5a68410d87#helix-gpt --rebuild
> > > helix-gpt> checking outputs of '/nix/store/z8d9k6fzax1ksv7phi23g6fg8ki3clrm-helix-gpt-0.34.drv'
> > > helix-gpt> Running phase: unpackPhase
> > > helix-gpt> unpacking source archive /nix/store/jnllq92xgrbfiyfnj9q4czlqkfi0ix68-source
> > > helix-gpt> source root is source
> > > helix-gpt> Running phase: patchPhase
> > > helix-gpt> Running phase: updateAutotoolsGnuConfigScriptsPhase
> > > helix-gpt> Running phase: installPhase
> > > helix-gpt> Running phase: fixupPhase
> > > helix-gpt> shrinking RPATHs of ELF executables and libraries in /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34
> > > helix-gpt> shrinking /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34/bin/helix-gpt
> > > helix-gpt> checking for references to /build/ in /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34...
> > > helix-gpt> patching script interpreter paths in /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34
> > > helix-gpt> stripping (with command strip and flags -S -p) in  /nix/store/mg7jpllxycw93p0pb9nb8q7krsybc5sm-helix-gpt-0.34/bin
> > > ┏━ Dependency Graph:
> > > ┃ ✔ helix-gpt-0.34
> > > ┣━━━ Builds
> > > ┗━ ∑ ⏵ 0 │ ✔ 1 │ ⏸ 0 │ Finished at 10:24:26 after 0s
> > > ```
> > 
> > that does not seem to use sandbox as nixpkgs-review did, I was expecting the same as here (pass on native Darwin, fail on sandbox)
> 
> I don't know why `nixpkgs-review-gha` explicitly writes `(Sandbox = true)` in its report. On Linux this is the default, so I was building `helix-gpt` with the sandbox enabled.

is it the default? then I don't know what causes the hash difference, cuz I built from Linux too and it worked (at https://github.com/NixOS/nixpkgs/pull/419604#issuecomment-3023544151)

repparw commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

GaetanLepage commented:
> I did `rm -rf ~/.cache/nixpkgs-review` before this one
> 
> I believe that should be enough, no?

No, this does not get rid of the relevant paths in the `/nix/store`.



> > OR the derivation is non-deterministic and for some reason it leads to two different hashes on our two machines.
> 
> Does running it again on gh actions gets you the same mismatched hash (i.e 2jkpSTUI... on x86)? Might be deterministic but depending on something else that is different?



> Does running it again on gh actions gets you the same mismatched hash (i.e 2jkpSTUI... on x86)?

Yes. I retried 5 mins ago:
```
 error: hash mismatch in fixed-output derivation '/nix/store/6sqdwbivsvf3gs5p5raw60ymn2xvbwd2-opencode-0.1.169-node-modules.drv':
         specified: sha256-TYjcA7MTfOKkvTwGSK1SotMDIH9xag79dikf6hMF8us=
            got:    sha256-2jkpSTUI/SBEFvVQDoGcXCe77beHupFP+BgU/K8PQ0M=
```



> Might be deterministic but depending on something else that is different?

Yes, this is probably what is happening. I can build it just fine on my `x86_64-linux` system.

GaetanLepage reviewed (approved):
Hydra was able to build `opencode` successfully. Let's merge then.

GaetanLepage commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,60 +1,147 @@
 
-  passthru.updateScript = nix-update-script { };
+    export MODELS_JSON="$(cat ${modelsDevData})"
+    bun build --define OPENCODE_VERSION="'${version}'" --compile --minify --target=${
+      bun-target.${stdenv.hostPlatform.system}
+    } --outfile=opencode ./packages/opencode/src/index.ts ${opencode-tui}/bin/opencode
```
```suggestion
    bun build \
      --define OPENCODE_VERSION="'${version}'" \
      --compile \
      --minify \
      --target=${bun-target.${stdenv.hostPlatform.system}} \
      --outfile=opencode
      ./packages/opencode/src/index.ts ${opencode-tui}/bin/opencode
```

Nit (can be ignored): This would be more readable.

GaetanLepage reviewed (commented):
Thanks for the great work @delafthi!

delafthi committed:
opencode: 0.1.69 -> 0.1.176

repparw commented:
space on pasted hashes on failing PR's

> I still have no success with fetchYarnDeps. For one, there is a bug with the generated yarn.lock file where @catalog: and @workspace:* are unquoted (see this https://github.com/oven-sh/bun/issues/20774 I opened). So, I have to manually fix the yarn.lock file.

Not sure, but seeing as current solution is working, I wouldn't put much time into the yarn solution.
Probably time is better spent into buildBunModule or the other bun hooks either way, as preference according to docs is to use the same runtime

delafthi commented:
> space on pasted hashes on failing PR's

thanks, fixed

delafthi commented:
> Not sure, but seeing as current solution is working, I wouldn't put much time into the yarn solution. Probably time is better spent into buildBunModule or the other bun hooks either way, as preference according to docs is to use the same runtime

I agree, `yarn.lock` and yarn hooks don't feel like a plug and play solution for this project.

repparw commented:
@GaetanLepage what did you get now? hash again?

GaetanLepage commented:
> @GaetanLepage what did you get now? hash again?

Yes, on my main system this time:
```
error: hash mismatch in fixed-output derivation '/nix/store/9dk86z2k04mj4i358ckg8izjfarvkzpg-opencode-0.1.176-node-modules.drv':
         specified: sha256-d2ux5d8hSKEOOEI0dA2biWTg46uMLkEb9uXK1A+kspc=
            got:    sha256-NRj6xzIV6OGNv46lKAoT5pIvQihIEq1t3oCtSKR9H/E=
```

delafthi commented:
> > @GaetanLepage what did you get now? hash again?
> 
> Yes, on my main system this time:
> 
> ```
> error: hash mismatch in fixed-output derivation '/nix/store/9dk86z2k04mj4i358ckg8izjfarvkzpg-opencode-0.1.176-node-modules.drv':
>          specified: sha256-d2ux5d8hSKEOOEI0dA2biWTg46uMLkEb9uXK1A+kspc=
>             got:    sha256-NRj6xzIV6OGNv46lKAoT5pIvQihIEq1t3oCtSKR9H/E=
> ```

I updated the hash. Though, this is kind of annoying.

Could that be because I'm crosscompiling? I'm working on `aarch64-darwin` and generating the hashes with `nix-build -A opencode --system x86_64-linux`.

delafthi commented:
> I don't think so, I'm native and it worked on mine. now probably it won't work on mine and work on gaetan's or viceversa haha. this is a weird one

Ah yes, I didn't see that 🙈

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`
Commit: `31defafa7d1a956e7251e244e5b5c192e4c83e8a`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

delafthi commented:
So, `--no-cache` is strictly necessary, else I get another hash after every run.

I reverted the hash back and added the `--force` flag to `bun install`. I'm not completely sure if that helps. The help sais

```
-f, --force    Always request the latest versions from the registry & reinstall all dependencies
```

delafthi commented:
Adding 

```
export HOME=$(mktemp -d)
export BUN_INSTALL_CACHE_DIR=$(mktemp -d)
```

allows to remove the `--no-cache` flag.

repparw commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`
Commit: `4721ddd76f0ebd076f1e782ad3d22ec21f6bffe2`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

GaetanLepage reviewed (approved):
Well done!

kirillrdy commented:
https://github.com/NixOS/nixpkgs/pull/419604/checks?check_run_id=45270188598

delafthi committed:
opencode: 0.1.176 -> 0.1.181

- Replaced `stdenv` with `stdenvNoCC`
- Moved subpackages into main derivation to enable updating subpackages with `nix-update-script`
- Fixed version test, by providing a temporary `HOME` directory

delafthi commented:
Added the following changes:
- Updated from 0.1.176 to 0.1.181
- Replaced `stdenv` with `stdenvNoCC`
- Moved subpackages into main derivation to enable updating subpackages with `nix-update-script`
- Fixed version test, by providing a temporary `HOME` directory

delafthi committed:
opencode: 0.1.181 -> 0.1.182

repparw commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`
Commit: `872fd0c102ca5f07a4bf140964f7f32ffdae6b20`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

repparw commented:
is there a nixpkgs policy about update frequency?

opencode releases a lot of versions, mostly small changes. should be update every x days at a maximum or something?

delafthi committed:
opencode: 0.1.182 -> 0.1.185

debashisbiswas commented:
On the current version, I'm getting the following build error on both x86_64-linux and aarch64-darwin:

```
error: builder for '/nix/store/cnc3w1jzjc5f57mnm2qpdva8bny6p83y-opencode-node_modules-0.1.185.drv' failed with exit code 1;
       last 11 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/px7ziy0kpds29mh0ih46ab9lagp14py8-source
       > source root is source
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: buildPhase
       > bun install v1.2.17 (282dda62)
       > Resolving dependencies
       > Resolved, downloaded and extracted [115]
       > error: lockfile had changes, but lockfile is frozen
       > note: try re-running without --frozen-lockfile and commit the updated lockfile
       For full logs, run:
         nix log /nix/store/cnc3w1jzjc5f57mnm2qpdva8bny6p83y-opencode-node_modules-0.1.185.drv
error (ignored): error: cannot unlink "/private/tmp/nix-build-opencode-tui-0.1.185-go-modules.drv-1/go/pkg/mod/github.com": Directory not empty
error: 1 dependencies of derivation '/nix/store/lywywsgbyx94s7ix05wxhv4n6qysrn1c-opencode-0.1.185.drv' failed to build
```

[As per the bun documentation](https://bun.sh/docs/cli/install):

> For reproducible installs, use --frozen-lockfile. This will install the exact versions of each package specified in the lockfile. **If your package.json disagrees with bun.lock, Bun will exit with an error. The lockfile will not be updated.**

I observed that in v0.1.185:
-  bun.lock includes the dependency `@types/bun` at [version `1.2.17`](https://github.com/sst/opencode/blob/v0.1.185/bun.lock#L56)
- opencode's package.json uses `@types/bun` at [version `latest`](https://github.com/sst/opencode/blob/v0.1.185/packages/opencode/package.json#L21
).

I did a `bun install` to create a new lockfile and compared it with the existing one: the new lockfile uses `1.2.18` while the old uses `1.2.17`. Perhaps this is a case where "your package.json disagrees with bun.lock", as package.json's `latest` in this case is now `1.2.18`, disagreeing with bun.lock's `1.2.17` and causing `--frozen-lockfile` to fail, but I'm not sure. This may also help explain the inconsistent `node_modules` hashes earlier in PR, though it doesn't seem like this dependency [has released new versions often enough](https://www.npmjs.com/package/@types/bun?activeTab=versions) to explain the frequency of that issue.

delafthi committed:
opencode: 0.1.182 -> 0.1.189

delafthi commented:
Yes, I have the same issue and it seems that there is no way to ignore that error and just build based on the lockfile.

delafthi committed:
opencode: 0.1.185 -> 0.1.194

delafthi commented:
> On the current version, I'm getting the following build error on both x86_64-linux and aarch64-darwin:
> 
> ```
> error: builder for '/nix/store/cnc3w1jzjc5f57mnm2qpdva8bny6p83y-opencode-node_modules-0.1.185.drv' failed with exit code 1;
>        last 11 log lines:
>        > Running phase: unpackPhase
>        > unpacking source archive /nix/store/px7ziy0kpds29mh0ih46ab9lagp14py8-source
>        > source root is source
>        > Running phase: patchPhase
>        > Running phase: updateAutotoolsGnuConfigScriptsPhase
>        > Running phase: buildPhase
>        > bun install v1.2.17 (282dda62)
>        > Resolving dependencies
>        > Resolved, downloaded and extracted [115]
>        > error: lockfile had changes, but lockfile is frozen
>        > note: try re-running without --frozen-lockfile and commit the updated lockfile
>        For full logs, run:
>          nix log /nix/store/cnc3w1jzjc5f57mnm2qpdva8bny6p83y-opencode-node_modules-0.1.185.drv
> error (ignored): error: cannot unlink "/private/tmp/nix-build-opencode-tui-0.1.185-go-modules.drv-1/go/pkg/mod/github.com": Directory not empty
> error: 1 dependencies of derivation '/nix/store/lywywsgbyx94s7ix05wxhv4n6qysrn1c-opencode-0.1.185.drv' failed to build
> ```
> 
> [As per the bun documentation](https://bun.sh/docs/cli/install):
> 
> > For reproducible installs, use --frozen-lockfile. This will install the exact versions of each package specified in the lockfile. **If your package.json disagrees with bun.lock, Bun will exit with an error. The lockfile will not be updated.**
> 
> I observed that in v0.1.185:
> 
>     * bun.lock includes the dependency `@types/bun` at [version `1.2.17`](https://github.com/sst/opencode/blob/v0.1.185/bun.lock?rgh-link-date=2025-07-04T07%3A09%3A08Z#L56)
> 
>     * opencode's package.json uses `@types/bun` at [version `latest`](https://github.com/sst/opencode/blob/v0.1.185/packages/opencode/package.json?rgh-link-date=2025-07-04T07%3A09%3A08Z#L21).
> 
> 
> I did a `bun install` to create a new lockfile and compared it with the existing one: the new lockfile uses `1.2.18` while the old uses `1.2.17`. Perhaps this is a case where "your package.json disagrees with bun.lock", as package.json's `latest` in this case is now `1.2.18`, disagreeing with bun.lock's `1.2.17` and causing `--frozen-lockfile` to fail, but I'm not sure. This may also help explain the inconsistent `node_modules` hashes earlier in PR, though it doesn't seem like this dependency [has released new versions often enough](https://www.npmjs.com/package/@types/bun?activeTab=versions) to explain the frequency of that issue.

Upstream updated the lockfile. Thus, this issue is fixed in 0.1.194.

repparw commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

GaetanLepage commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,60 +1,182 @@
+
+    dontConfigure = true;
+
+    buildPhase = ''
```
Missing `runHook {pre,post}Build`.

delafthi replied:
Done

GaetanLepage commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,60 +1,182 @@
+
+    buildPhase = ''
+
+      export HOME=$(mktemp -d)
```
```suggestion
```

Add `writableTmpDirAsHomeHook` to `nativeBuildInputs` instead.

delafthi replied:
Done

GaetanLepage commented on pkgs/by-name/op/opencode/package.nix:
```diff
@@ -1,60 +1,188 @@
+    env.CGO_ENABLED = 0;
+
+    ldflags = [
+      "-s"
+      "-w"
```
```suggestion
      "-s"
```
It seems that `-w` is now redundant: https://tip.golang.org/doc/go1.22#linker

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419604`
Commit: `4acc22caf616f5958042d2746921f39203277edc`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>opencode</li>
  </ul>
</details>

GaetanLepage reviewed (approved):
Good job everyone.
